site/app/libraries/MultiIterator.php

```diff
@@ -2,27 +2,38 @@
 
 namespace app\libraries;
 
+use app\exceptions\NotImplementedException;
+
 /**
  * Class MultiIterator
  * @package app\libraries
  *
- * Allows the chaining of multiple iterators in one iterator object
+ * An iterator that allows for sequentially iterating over each passed
+ * iterator in-order, one at a time. This means that iterator2 will not be
+ * called until iterator1 has been exhausted. This differs from
+ * MultipleIterator as that commits all iterator functions over all attached
+ * iterators in parallel (i.e. calling next() will return an array with the
+ * results of next() on all attached iterators).
+ *
+ * @see \MultipleIterator
  */
 class MultiIterator implements \Iterator {
 
-    /** @var array array of [\Closure, $this] */
-    private $iterator_generators = [];
+    /** @var \Iterator[] */
+    private $iterators = [];
+    /** @var int */
+    private $iterator_key = 0;
     /** @var \Iterator */
     private $curr_it = null;
```

Comment:
We can get rid of the `curr_it` field since we're using `iterator_key` to index the array of iterators